### PR TITLE
Flax Remat for LongT5

### PR DESCRIPTION
**Pull request by KMFODA**
# What does this PR do?

Following on from the discussion in https://github.com/huggingface/transformers/issues/17399 this PR adds remat (flax's gradient checkpointing) to LongT5. Most of the code is copied from https://github.com/huggingface/transformers/pull/17843/files and then altered to be applicable for LongT5. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@sanchit-gandhi or @patrickvonplaten

### CODE REVIEW

**Comment by sanchit-gandhi** [comment]
Hey @KMFODA, looks like a great start on this PR! 

What remains to do is wrapping the Flax T5 Block Collection in the `remat` transformation when `gradient_checkpointing` is passed as `True`. This is analogous to wrapping the Flax Bert Layer Collection in the `remat` operation in the reference PR https://github.com/huggingface/transformers/pull/17843.

We just need to update the following lines:
https://github.com/huggingface/transformers/blob/f0982682bd6fd0b438dda79ec45f3a8fac83a985/src/transformers/models/t5/modeling_flax_t5.py#L648-L651

To reflect what is done in Flax Bert:
https://github.com/huggingface/transformers/blob/f0982682bd6fd0b438dda79ec45f3a8fac83a985/src/transformers/models/bert/modeling_flax_bert.py#L553-L558

As a pointer, the `static_argnums` kwarg is used to denote any boolean arguments passed to the `remat` layer. Inspecting the `__call__` method of FlaxT5LayerCollection, we see that the final four arguments are booleans:
https://github.com/huggingface/transformers/blob/f0982682bd6fd0b438dda79ec45f3a8fac83a985/src/transformers/models/t5/modeling_flax_t5.py#L616-L628
Thus, we have `static_argnums=(6,7,8,9)` (corresponding to `output_attentions`, `return_dict`, `deterministic` and `init_cache`).

**Comment by KMFODA** [comment]
Hey @sanchit-gandhi. Sorry this is taking so long, adding your changes was relatively easy but I'm a bit stuck trying to pass a few failing tests which I believe are caused by my addition of `key_value_states`, `encoder_hidden_states`, `encoder_attention_mask`, `use_cache` and `init_cache` into most calls as I remember in your PR you mentioned that remat doesn't take kwargs.

I believe somewhere in the code `key_value_states` is being incorrectly set to a non `None` figure which is causing some of these errors but I'm not sure where.

**Review by sanchit-gandhi — COMMENTED** [review]
Hey @KMFODA! Sorry, I think there was some confusion with the kwarg comment. It's just in the layer that we transform through the `remat` (gradient checkpointing) operation where we have to omit kwargs (see https://github.com/huggingface/transformers/blob/1763770bd9e0d487cc12fe0624abd30a9c3b4cfb/src/transformers/models/bert/modeling_flax_bert.py). In every other `__call__` method, kwargs are perfectly valid. 

For this PR, could we stick to just adding gradient checkpointing? We can open a second PR that introduces the `key_value_states` and `use_cache` args. Adding both at in one go makes it harder to test what is/isn't working! In this case, is the issue with gradient checkpointing? Or with the `key_value_states`? If we tackle them in separate PRs, we'll be able to iterate more efficiently!

**Comment by KMFODA** [comment]
of course that makes sense. Apologies for the misunderstanding. I'll work on the gradient checkpointing part using your suggestions and remove the `key_value_states` for this PR.

**Comment by KMFODA** [comment]
hey @sanchit-gandhi. I beleive this is now ready for review. The PR passes all the tests except ones related to inconsistencies between t5 and long_t5. If you're happy with this let me know and then I think running `make fix-copies` should fix the remaining errors?

File: src/transformers/models/t5/modeling_flax_t5.py
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:687]: Note that we don't actually need to propagate `return_dict` to the `FlaxT5LayerCollection` - this arg is not used by the `FlaxT5LayerCollection` or `FlaxT5Layer`. Thus, it can be removed from here and from the `layer_module call`, and the `static_argnums` updated accordingly.

File: src/transformers/models/t5/modeling_flax_t5.py
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:1333]: Once the aforementioned `return_dict` changes are made to the `FlaxT5BlockCollection` this change won't be necessary :)

File: src/transformers/models/t5/modeling_flax_t5.py
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:1443]: And again! Not necessary once the `return_dict` kwarg is removed from the `layer_module` call in `FlaxT5BlockCollection`

File: src/transformers/models/t5/modeling_flax_t5.py
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:710]: `return_dict` not in fact used by the `FlaxT5LayerCollection` - we can omit this arg (as was done previously)!

**Review by sanchit-gandhi — COMMENTED** [review]
Hey @KMFODA, this looks good with regards to the gradient checkpointing! It's pretty fiddly to get working, but we're almost there. 

Going through the PR, I noticed that the `FlaxT5LayerCollection` accepts `return_dict` as a kwarg, but actually does not use it in it's `__call__` method. To make our lives easier, we can omit this argument when we call the `FlaxT5LayerCollection` (as was done previously in the `FlaxT5BlockCollection`). This will then mean the change of setting `return_dict: bool = True` can be undone, and set back to `return_dict: None` in the downstream class.

After this quick change has been implemented, feel free to run `make fix-copies` to duplicate remat across the relevant Flax T5 inherited modules. I can then take a final look and get this approved!

**Comment by KMFODA** [comment]
Thanks @sanchit-gandhi for all the helpful comments. Addressed them all and ran `make fix-copies`. Hopefully these changes should be reflected properly for `LongT5 `as well now.

File: tests/models/t5/test_modeling_flax_t5.py
Comment (sanchit-gandhi) [inline tests/models/t5/test_modeling_flax_t5.py:247]: Great that this test passes in the standalone test class! In the final version of the PR #17843, the `test_gradient_checkpointing` test was moved to `test_modeling_flax_common`:
https://github.com/huggingface/transformers/blob/ab2006e3d6db88654526a4169e65d4bfc52da2e3/tests/test_modeling_flax_common.py#L1102
So this test will be run automatically for all modules that have the `.enable_gradient_checkpointing()` method implemented (in this case FlaxT5 and FlaxLongT5). This means we can remove this test from the `test_modeling_flax_t5` file!

**Review by sanchit-gandhi — APPROVED** [review]
Thanks @KMFODA for adding `remat` for T5 and LongT5! It's certainly not an easy job 😅 Hopefully this allows you to bump up your batch sizes while training!

The PR looks good to me! Requesting a second look from @patil-suraj to verify all is well before we merge this!

File: tests/models/t5/test_modeling_flax_t5.py
Comment (KMFODA) [reply tests/models/t5/test_modeling_flax_t5.py]: ah good to know.  I've removed it now.

Comment (sanchit-gandhi) [reply tests/models/t5/test_modeling_flax_t5.py]: Brill, thanks! PR is good to go once we get the green light from @patil-suraj!

**Comment by sanchit-gandhi** [comment]
Thanks for this PR @KMFODA! Awesome stuff, excited to hear how you get on with improved memory usage with the aid of checkpointing!

**Comment by KMFODA** [comment]
Amazing, will keep you posted. Thanks for all the help getting this merged!

**Pull request merged**
